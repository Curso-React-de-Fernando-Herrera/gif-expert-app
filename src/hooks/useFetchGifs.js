import { useEffect, useState } from "react"
import { GetGifs } from 'helpers/getGifs'


export const useFetchGifs = ({ category }) => {
  const INITIAL_STATE = {
    gifs: [],
    loading: true
  }

  const [state, setState] = useState(INITIAL_STATE)

  useEffect(() => {

    GetGifs({ category })
      .then(gifs => {

        setState({
          gifs,
          loading: false
        })

      })

  }, [category])

  return state
}