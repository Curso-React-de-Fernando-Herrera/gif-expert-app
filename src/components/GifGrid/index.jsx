import GifGridItem from 'components/GifGridItem'
import { useFetchGifs } from 'hooks/useFetchGifs'
import PropTypes from 'prop-types'

const GifGrid = ({ category }) => {
  const { gifs, loading } = useFetchGifs({ category })

  return <>
    <h2>{category}</h2>

    { loading && <p>Loading...</p> }
    
    <div>
      {
        gifs.map(gif => 
          <GifGridItem
            key={gif.id}
            title={gif.title}
            username={gif.username}
            url={gif.url}
          />
        )
      }
    </div>
  </>
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}

export default GifGrid
