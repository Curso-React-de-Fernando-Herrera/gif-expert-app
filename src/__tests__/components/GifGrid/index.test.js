import { shallow } from 'enzyme'
import GifGrid from 'components/GifGrid'
import { useFetchGifs } from 'hooks/useFetchGifs'

jest.mock('hooks/useFetchGifs') // simula una ejecucion a ese archivo y controla que va a devolver

describe('<GifGrid /> component', () => {
  useFetchGifs.mockReturnValue({
    gifs: [],
    loading: true
  })

  const DEFAULT_GIF = [{
    id: 'abc',
    title: 'examle',
    url: 'https://localhost/example.png',
    username: 'pepito'
  }]

  let category = 'Dragon Ball'
  let wrapper = shallow(<GifGrid category={category} />)

  beforeEach(() => {
    useFetchGifs.mockReturnValue({
      gifs: [],
      loading: true
    })
    wrapper = shallow(<GifGrid category={category} />)
  })

  test('should be render the loading at first time', () => {
    const initialLoadingText = 'Loading...'
    const loadingText = wrapper.find('p').text().trim()

    expect(wrapper).toMatchSnapshot()
    expect(loadingText).toBe(initialLoadingText)
  })

  test('should be return gif data', () => {
    useFetchGifs.mockReturnValue({
      gifs: DEFAULT_GIF,
      loading: false
    })

    wrapper = shallow(<GifGrid category={category} />)
    expect(wrapper).toMatchSnapshot()
  })
})