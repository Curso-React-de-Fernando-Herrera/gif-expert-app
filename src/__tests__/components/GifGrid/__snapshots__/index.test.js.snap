// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GifGrid /> component should be render the loading at first time 1`] = `
<Fragment>
  <h2>
    Dragon Ball
  </h2>
  <p>
    Loading...
  </p>
  <div />
</Fragment>
`;

exports[`<GifGrid /> component should be return gif data 1`] = `
<Fragment>
  <h2>
    Dragon Ball
  </h2>
  <div>
    <GifGridItem
      key="abc"
      title="examle"
      url="https://localhost/example.png"
      username="pepito"
    />
  </div>
</Fragment>
`;
